lib/bolt/transport/docker/connection.rb

```diff
@@ -29,203 +48,113 @@ def connect
           true
         rescue StandardError => e
           raise Bolt::Node::ConnectError.new(
-            "Failed to connect to #{@target.safe_name}: #{e.message}",
+            "Failed to connect to #{target.safe_name}: #{e.message}",
             'CONNECT_ERROR'
           )
         end
 
-        # Executes a command inside the target container
+        # Executes a command inside the target container. This is called from the shell class.
         #
-        # @param command [Array] The command to run, expressed as an array of strings
-        # @param options [Hash] command specific options
-        # @option opts [String] :interpreter statements that are prefixed to the command e.g `/bin/bash` or `cmd.exe /c`
-        # @option opts [Hash] :environment A hash of environment variables that will be injected into the command
-        # @option opts [IO] :stdin An IO object that will be used to redirect STDIN for the docker command
-        def execute(*command, options)
-          command.unshift(options[:interpreter]) if options[:interpreter]
-          # Build the `--env` parameters
-          envs = []
-          if options[:environment]
-            options[:environment].each { |env, val| envs.concat(['--env', "#{env}=#{val}"]) }
+        # @param command [string] The command to run
+        def execute(command)
+          env_vars = []
+          # CODEREVIEW: Is there a better way to handle `docker exec ... PT_foo` not working well?
+          if command.start_with?("PT_")
+            parts = Shellwords.split(command)
+            env_vars, command = parts.partition { |p| p.start_with?("PT_") }
+            command = Shellwords.shelljoin(command)
           end
 
-          command_options = []
-          # Need to be interactive if redirecting STDIN
-          command_options << '--interactive' unless options[:stdin].nil?
-          command_options << '--tty' if options[:tty]
-          command_options.concat(envs) unless envs.empty?
-          command_options << container_id
-          command_options.concat(command)
+          args = []
+          # CODEREVIEW: Is it always safe to pass --interactive?
+          args += %w[--interactive]
```

Comment:
It doesn't seem like it'd be an issue. If we're really concerned about always keeping STDIN open, we could either pass `options` to this method or handle building the command a different way (see above comment).

Reply:
Yeah, I think similar to above it's hard to know if any other transports will need to know about 'options' or not, and if *this* was the only thing that ended up needing it that'd make me sad to have to pass them around and accept them *everywhere* just for this. Unless you (or anyone else) is concerned I think we can leave it here for now and can change it if it causes issues or if other transports need to know about options.